content/en/blog/_posts/2024-09-20-kubeadm-v1beta4.md

```diff
@@ -0,0 +1,100 @@
+---
+layout: blog
+title: 'Kubernetes v.31: kubeadm v1beta4'
+date: 2024-09-20
+slug: kubernetes-1-31-kubeadm-v1beta4
+author: >
+   Paco Xu (DaoCloud),
+---
+
+[`kubeadm`](/docs/reference/setup-tools/kubeadm/) Configuration([v1beta4](https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta4/)) is released with Kubernetes v1.31.
+The configuration 
+v1beta3 is still supported, but v1beta3 will be deprecated and removed in future releases.
+
+This guide will walk you through key changes and update about kubeadm v1beta4 Configuration,
+and how to migrate from v1beta3 to v1beta4.
+
+kubeadm v1beta4 configuration details can be found in the [API doc]((https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta4/)).
+The documentation is maintained in [code](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go) and generated to the website kubernetes.io.
+
+### A list of changes since v1beta3
+
+This version improves on the [v1beta3](https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta3/)
+ format by fixing some minor issues and adding a few new fields.
+
+To put it simply,
+
+- Two new API type: ResetConfiguration and UpgradeConfiguration
+- For InitConfiguration and JoinConfiguration, `dry-run` mode and `NodeRegistration.ImagePullSerial` are supported
+- For ClusterConfiguration, we add field including `CertificateValidityPeriod`, `CACertificateValidityPeriod`, `EncryptionAlgorithm`, `DNS.Disabled` and `Proxy.Disabled`.
+- Support `ExtraEnvs` for all control plan components
+- `ExtraArgs` changed from a map to structured extra arguments for duplicates
+- Add a `Timeouts` structure for init, join, upgrade and reset.
```

Comment:
Please always use the user-facing field names rather than the Go field names when possible. For example, say "imagePullSerial" instead of "ImagePullSerial".

Reply:
/cc @neolit123 
Do we want to change the style here? 

In https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta3/, the changes docs are using go fields like the release notes.

In https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta3/#resource-types and below, we use the user-facing field names.

Reply:
For clarification's purpose. The config API reference is NOT generated directly from the k/k repository for two reasons:

- The package in k/k/cmd/kubeadm is not importable from an external project.
- The comments in the source code are not documenting fields in the way users see them.

What you are reading from k/website is generated from github.com/tengqm/kubeconfig repo, where I manually copied the kubeadm api repository and manually synchronize whatever changes made to k/k/cmd/kubeadm/api.

The "changes" doc are using Go fields because I failed to perform a line-by-line checking when that API was introduced.

Reply:
> Please always use the user-facing field names rather than the Go field names when possible. For example, say "imagePullSerial" instead of "ImagePullSerial".

i don't mind this format of using the JSON tags. this is something the docs team has been recommending to us on the blog and web pages and i agree with it.

> Do we want to change the style here?

however for the style in the go.doc, that's also targeting developers, so i think field names also work.
we could make it simpler to use JSON field names everywhere. but that's like something we can do after 1.31 or another time.

Reply:
I add https://github.com/kubernetes/website/pull/47248/commits/41b8f0548c7822e1be2e2e1021221303a350dadc to make most places using user facing field names.